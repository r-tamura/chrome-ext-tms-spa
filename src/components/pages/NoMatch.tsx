import * as React from "react";

export const NoMatch: React.FC = () => {
  return (
    <main className="main">
      <h1>Not Found 404</h1>
    </main>
  );
};
