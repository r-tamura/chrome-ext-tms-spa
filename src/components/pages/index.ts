export * from "./TimeCardPage";
export * from "./SignInPage";
export * from "./TransportExpensePage";
export * from "./NoMatch";
export * from "./App";
export * from "./Root";
