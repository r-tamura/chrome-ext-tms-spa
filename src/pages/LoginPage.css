.login-center {
  align-self: center;

  & > * {
    text-align: center;
  }
}

