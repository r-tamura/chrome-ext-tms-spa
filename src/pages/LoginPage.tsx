import * as React from "react"
import { Helmet } from "react-helmet"
import { RouteComponentProps } from "react-router-dom"
import { Form, Text } from "react-form"
import { RootState } from "~/modules"
import { loginUser, navigateToDashBoard } from "~/modules/user"
import { Dispatch } from "redux"
import { connect } from "react-redux"

interface IProps extends React.Props<{}>, RouteComponentProps<{}> {
  readonly isFetching: boolean,
  /* ログイン済みであるか */
  readonly isAuthenticated: boolean,
  /* ユーザー名 */
  readonly name: string,
  loginUser: (id: string, pw: string) => any
  navigateToDashBoard: () => any
}

/**
 * ログインページ
 */
class LoginPage extends React.Component<IProps, {}> {

  public componentWillMount() {
    this.redirectToDashboardIfNeeded(this.props.isAuthenticated)
  }

  public render() {
    const { isFetching } = this.props
    if (isFetching) {
      return <p>Fetching...</p>
    }

    return (
      <main className="main main--full tms-grid-3-container">
        <Helmet>
          <title>Login | TMS</title>
        </Helmet>
        <div className="login-main tms-grid--offset1 tms-grid--col1 tms-grid-sm--col3">
          <h1>Sign in to TMS</h1>
          <p>Enter your <strong>user name</strong> and <strong>password</strong>.</p>
          <Form onSubmit={this.handleSubmit}>
            { formApi => (
              <div className="tms-panel">
                <form onSubmit={formApi.submitForm}>
                  {/* User name text filed */}
                  <div className="tms-textfield">
                    <Text
                      field={"username"}
                      id={"username"}
                      placeholder="John Doe"
                      maxLength={30}
                      required={true}
                    />
                    <label htmlFor={"username"}>UserName</label>
                  </div>
                  {/* Password text field */}
                  <div className="tms-textfield">
                    <Text
                      type="password"
                      field={"password"}
                      id={"password"}
                      placeholder="Your password"
                      maxLength={30}
                      required={true}
                    />
                    <label htmlFor={"password"}>Password</label>
                  </div>
                  <button className="tms-btn primary block">SIGN IN</button>
                </form>
              </div>
            )}
          </Form>
        </div>
      </main>
    )
  }

  private shouldRedirectToDashBoard(isAuthenticated: boolean): boolean {
    return isAuthenticated
  }

  private redirectToDashboardIfNeeded(isAuthenticated: boolean) {
    if (this.shouldRedirectToDashBoard(isAuthenticated)) {
      this.props.navigateToDashBoard()
    }
  }

  private handleSubmit = ({username = "", password = ""}) => {
    this.props.loginUser(username, password)
  }
}

const mapStateToProps = (state: RootState, props: IProps) => {
  const { isFetching, isAuthenticated, name } = state.user
  return {
    isFetching,
    isAuthenticated,
    name,
  }
}

const mapDispatchToProps = {
  loginUser,
  navigateToDashBoard,
}

export {
  LoginPage,
}

// export default withNav<RouteComponentProps<{}>>(Login)
export default connect(mapStateToProps, mapDispatchToProps)(LoginPage)
