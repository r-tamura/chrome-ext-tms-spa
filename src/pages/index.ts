export * from "./Attendance";
export * from "./SignInPage";
export * from "./TransportExpensePage";
export * from "./App";
export * from "./NoMatch";
