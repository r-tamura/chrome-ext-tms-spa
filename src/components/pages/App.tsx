import React from "react";
import { Helmet } from "react-helmet";
import { Route, Switch, Redirect } from "react-router-dom";
import { AppHeader, Nav } from "~/components/organisms";
import { EnsureLoggedInContainer } from "~/components/containers/EnsureLoggedInContainer";
import { TransportExpensePage, TimeCardPage, SignInPage, NoMatch } from ".";
import { AppGrid } from "~/components/organisms";

const PAGES_WITH_NAVBAR = ["/transportation", "/attendance"];

export function App() {
  return (
    <AppGrid>
      <Helmet>
        <link
          href="https://fonts.googleapis.com/css?family=M+PLUS+Rounded+1c"
          rel="stylesheet"
        ></link>
        <meta
          name="viewport"
          content="width=device-width, initial-scale=1.0, shrink-to-fit=no"
        />
      </Helmet>
      <AppHeader />
      <Route
        path={PAGES_WITH_NAVBAR}
        render={renderProps => {
          return <Nav path={renderProps.match.url} />;
        }}
      />
      <Switch>
        <Route
          exact={true}
          path="/"
          component={() => <Redirect to={"/signin"} />}
        />
        <Route exact={true} path="/signin" component={SignInPage} />
        <EnsureLoggedInContainer>
          {/* TODO: Dashboard画面作成までは交通費管理画面へリダイレクト */}
          <Route
            path="/dashboard"
            component={() => <Redirect to={"/transportation"} />}
          />
          <Route path="/transportation" component={TransportExpensePage} />
          <Route path="/attendance" component={TimeCardPage} />
        </EnsureLoggedInContainer>
        <Route component={NoMatch} />
      </Switch>
    </AppGrid>
  );
}
