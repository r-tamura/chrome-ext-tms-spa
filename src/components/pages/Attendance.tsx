import React from "react";
import styled from "styled-components";
import { Helmet } from "react-helmet";
import { SelectBox } from "~/components/SelectBox";
import { Button, ButtonGroup } from "~/components/atoms";
import { zerofill } from "~/helpers/common";
import { AttendanceMonthlyView } from "~/types";
import { MsgAttendance } from "~/helpers/_const";
import { useAttendance } from "~/stores/hooks";
import { MainContent } from "~/components/organisms";
import { BOLD } from "~/styles/font";

export const AttendancePage: React.FC = () => {
  const {
    projects,
    fetchAttendances,
    attendanceSettings,
    updateSettings,
    attendanceMonthly,
    saveAttendances,
    updateDaily,
    changeMonth,
    setMonthlyWithDefaults
  } = useAttendance();

  const { year, month } = attendanceMonthly;
  return (
    <main className="app-content-with-navbar">
      <Helmet>
        <title>Attendance | TMS</title>
      </Helmet>
      <MainContent>
        <h1>
          <Button id={"btn-prev-month"} onClick={onPrevMonthClick}>
            {"<"}
          </Button>
          {year} / {month}
          <Button id={"btn-next-month"} onClick={onNextMonthClick}>
            {">"}
          </Button>
        </h1>
        {renderMonthly(attendanceMonthly)}

        {/* 時間リスト */}
        <datalist id="timelist-am">
          <option>09:00</option>
          <option>09:30</option>
          <option>10:00</option>
          <option>10:30</option>
        </datalist>
        <datalist id="timelist-pm">
          <option>17:30</option>
          <option>18:00</option>
          <option>18:30</option>
          <option>19:00</option>
          <option>19:30</option>
          <option>19:30</option>
          <option>20:00</option>
          <option>20:30</option>
          <option>21:00</option>
          <option>21:30</option>
          <option>22:00</option>
          <option>22:30</option>
          <option>23:00</option>
          <option>23:30</option>
        </datalist>
      </MainContent>
    </main>
  );

  function renderMonthly({
    days,
    isFetching,
    hasApplied
  }: AttendanceMonthlyView) {
    if (isFetching) {
      // TODO: ロードインジケータを表示するように
      return <p>Now Loading...</p>;
    }

    if (days.length === 0) {
      return <p>勤怠データを取得できません</p>;
    }

    return (
      <div>
        <ButtonGroup>
          <Button id={"btn-set-default"} onClick={onSetDefaultClick}>
            set default
          </Button>
          <Button id={"btn-save"} onClick={onSaveClick}>
            Save
          </Button>
          <Button id={"btn-fetch"} onClick={fetchAttendances}>
            Reload
          </Button>
          {hasApplied ? (
            <Button id={"btn-submit"} title={"上長申請中"} disabled>
              Submit
            </Button>
          ) : (
            <Button id={"btn-submit"} title={"上長申請"} onClick={onSubmit}>
              Submit
            </Button>
          )}
        </ButtonGroup>
        <div>
          {/* <SelectBox
            value={attendanceSettings.projectId || ""}
            options={{
              items: projects,
              valueKey: "projectId",
              labelKey: "name"
            }}
            onChange={$select =>
              updateSettings({
                projectId: $select.selectedOptions[0].value
              })
            }
          /> */}
          select box
        </div>
        <table className="tms-table tms-table--bordered">
          <thead>
            <tr>
              <th>日</th>
              <th>出勤</th>
              <th>退勤</th>
              {/* <th>合計</th>
              <th>Extra</th>
              <th>MN Extra</th> */}
              <th>プロジェクト</th>
              <th>承認</th>
            </tr>
          </thead>
          <tbody>
            {days.map((a, index) => (
              <tr
                key={a.dailyId}
                className={a.hasUpdated ? "attendance-diff" : ""}
              >
                {/* <td
                  className={
                    a.isWeekday ? "attendance-weekday" : "attendance-holiday"
                  }
                >
                  {zerofill(2, index + 1)}
                </td> */}
                {a.isWeekday ? (
                  <TdWeekday>{zerofill(2, index + 1)}</TdWeekday>
                ) : (
                  <TdHoliday>{zerofill(2, index + 1)}</TdHoliday>
                )}
                <td className="tms-textfield tms-textfield--table">
                  <input
                    type="time"
                    value={a.start}
                    list="timelist-am"
                    onChange={e =>
                      updateDaily(a.dailyId, {
                        start: e.target.value
                      })
                    }
                  />
                </td>
                <td className="tms-textfield tms-textfield--table">
                  <input
                    type="time"
                    value={a.end}
                    list="timelist-pm"
                    onChange={e =>
                      updateDaily(a.dailyId, { end: e.target.value })
                    }
                  />
                </td>
                {/* プロジェクトが未設定の場合は表示なし */}
                <td>
                  {/* <SelectBox
                    value={a.project ? a.project.projectId : ""}
                    options={{
                      items: projects,
                      valueKey: "projectId",
                      labelKey: "name"
                    }}
                    onChange={$select =>
                      updateDaily(a.dailyId, {
                        projectId: $select.selectedOptions[0].value
                      })
                    }
                  // additionalClass={"tms-select--table"}
                  /> */}
                  select box
                </td>
                <td>
                  {a.hasConfirmed
                    ? MsgAttendance.HAS_SUBMITED
                    : MsgAttendance.NOT_SUBMITED}
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    );
  }

  function onSaveClick(): void {
    saveAttendances();
  }

  function onSetDefaultClick(): void {
    setMonthlyWithDefaults(attendanceMonthly.days.map(d => d.dailyId));
  }

  function onNextMonthClick(): void {
    const { year: curYear, month: curMonth } = attendanceMonthly;
    const nextMonth = (curMonth % 12) + 1;
    const nextYear = nextMonth === 1 ? curYear + 1 : curYear;
    changeMonth(nextYear, nextMonth);
  }

  function onPrevMonthClick(): void {
    const { year: curYear, month: curMonth } = attendanceMonthly;
    const prevMonth = ((curMonth + 10) % 12) + 1;
    const prevYear = prevMonth === 12 ? curYear - 1 : curYear;
    changeMonth(prevYear, prevMonth);
  }

  function onSubmit(): void {
    console.log("[Attendance] Submitted");
  }
};

const TdHoliday = styled.td`
  color: rgba(255, 0, 0, 0.4);
  font-weight: ${BOLD};
`;

const TdWeekday = styled.td``;
