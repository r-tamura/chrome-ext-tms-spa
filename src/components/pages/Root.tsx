import React from "react";
import { Store, AnyAction } from "redux";
import { Provider as ReduxProvider } from "react-redux";
import { ConnectedRouter } from "connected-react-router";
import { ThemeProvider } from "styled-components";
import { history } from "~/stores";
import { App } from "~/components/pages";
import { RootState } from "~/modules";
import { telema } from "~/styles/theme";
import { GlobalStyle } from "~/styles/GlobalStyle";

interface Props {
  store: Store<RootState, AnyAction>;
}

export const Root: React.FC<Props> = ({ store }) => {
  return (
    <ReduxProvider store={store}>
      <ThemeProvider theme={telema}>
        <ConnectedRouter history={history}>
          <>
            <App />
            <GlobalStyle />
          </>
        </ConnectedRouter>
      </ThemeProvider>
    </ReduxProvider>
  );
};
