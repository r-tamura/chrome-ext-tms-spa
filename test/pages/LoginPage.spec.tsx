import React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import { LoginPage } from "~/pages/LoginPage";
import { getMockRouterProps } from "../__mocks__/";

describe("<LoginPage />", () => {
  const mockRoute = getMockRouterProps(null);
  it("should render 'fetching...'", () => {
    const mockFunc1 = jest.fn();
    const mockFunc2 = jest.fn();
    const page = renderer
      .create(
        <LoginPage
          isFetching={true}
          isAuthenticated={false}
          name={""}
          loginUser={mockFunc1}
          navigateToDashBoard={mockFunc2}
          {...mockRoute}
        />
      )
      .toJSON();
    expect(page).toMatchSnapshot();
  });

  it("should render Login page", () => {
    const mockFunc1 = jest.fn();
    const mockFunc2 = jest.fn();
    const component = mount(
      <LoginPage
        isFetching={false}
        isAuthenticated={false}
        name={"test"}
        loginUser={mockFunc1}
        navigateToDashBoard={mockFunc2}
        {...mockRoute}
      />
    );
    component.find("form").simulate("submit", { preventDefault: () => ({}) });
    expect(toJson(component)).toMatchSnapshot();
  });

  it("should redirect to DashBoard", () => {
    const mockFunc1 = jest.fn();
    const mockFunc2 = jest.fn();
    const component = mount(
      <LoginPage
        isFetching={false}
        isAuthenticated={true}
        name={"test"}
        loginUser={mockFunc1}
        navigateToDashBoard={mockFunc2}
        {...mockRoute}
      />
    );
    expect(mockFunc2).toBeCalled();
  });
});
