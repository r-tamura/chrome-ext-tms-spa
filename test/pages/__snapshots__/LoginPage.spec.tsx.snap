// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginPage /> should render 'fetching...' 1`] = `
<p>
  Fetching...
</p>
`;

exports[`<LoginPage /> should render Login page 1`] = `
<LoginPage
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 2,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "",
        "pathname": "",
        "search": "",
        "state": Object {},
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  isAuthenticated={false}
  isFetching={false}
  location={
    Object {
      "hash": "",
      "key": "",
      "pathname": "",
      "search": "",
      "state": Object {},
    }
  }
  loginUser={[MockFunction]}
  match={
    Object {
      "isExact": true,
      "params": null,
      "path": "",
      "url": "",
    }
  }
  name="test"
  navigateToDashBoard={[MockFunction]}
  staticContext={Object {}}
>
  <div
    className="main main--full tms-grid-12-col-parent"
  >
    <HelmetWrapper
      defer={true}
      encodeSpecialCharacters={true}
    >
      <SideEffect(NullComponent)
        defer={true}
        encodeSpecialCharacters={true}
        title="Login | TMS"
        titleAttributes={Object {}}
      >
        <NullComponent
          defer={true}
          encodeSpecialCharacters={true}
          title="Login | TMS"
          titleAttributes={Object {}}
        />
      </SideEffect(NullComponent)>
    </HelmetWrapper>
    <div
      className="tms-grid--offset3 tms-grid--col7 login-center"
    >
      <h1>
        Sign in to TMS
      </h1>
      <p>
        Enter your 
        <strong>
          user name
        </strong>
         and 
        <strong>
          password
        </strong>
        .
      </p>
      <t
        onSubmit={[Function]}
      >
        <Connect(t)
          onSubmit={[Function]}
          store={
            Object {
              "dispatch": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
              Symbol(observable): [Function],
            }
          }
        >
          <t
            dispatch={[Function]}
            formState={
              Object {
                "asyncErrors": Object {},
                "asyncSuccesses": Object {},
                "asyncValidations": 0,
                "asyncWarnings": Object {},
                "errors": Object {},
                "submits": 1,
                "submitted": false,
                "submitting": true,
                "successes": Object {},
                "touched": Object {
                  "password": true,
                  "username": true,
                },
                "validating": Object {},
                "validationFailed": Object {},
                "validationFailures": 0,
                "values": Object {},
                "warnings": Object {},
              }
            }
            onSubmit={[Function]}
            store={
              Object {
                "dispatch": [Function],
                "getState": [Function],
                "replaceReducer": [Function],
                "subscribe": [Function],
                Symbol(observable): [Function],
              }
            }
            storeSubscription={
              Subscription {
                "listeners": Object {
                  "clear": [Function],
                  "get": [Function],
                  "notify": [Function],
                  "subscribe": [Function],
                },
                "onStateChange": [Function],
                "parentSub": undefined,
                "store": Object {
                  "dispatch": [Function],
                  "getState": [Function],
                  "replaceReducer": [Function],
                  "subscribe": [Function],
                  Symbol(observable): [Function],
                },
                "unsubscribe": [Function],
              }
            }
          >
            <div
              className="tms-panel"
            >
              <form
                onSubmit={[Function]}
              >
                <div
                  className="tms-textfield"
                >
                  <r
                    field="username"
                    id="username"
                    maxLength={30}
                    placeholder="John Doe"
                    required={true}
                  >
                    <t
                      fieldApi={
                        Object {
                          "doneValidatingField": [Function],
                          "getError": [Function],
                          "getFieldName": [Function],
                          "getSuccess": [Function],
                          "getTouched": [Function],
                          "getValue": [Function],
                          "getWarning": [Function],
                          "registerAsyncValidation": [Function],
                          "reset": [Function],
                          "setError": [Function],
                          "setSuccess": [Function],
                          "setTouched": [Function],
                          "setValue": [Function],
                          "setWarning": [Function],
                          "submits": 1,
                          "submitted": false,
                          "validatingField": [Function],
                        }
                      }
                      id="username"
                      maxLength={30}
                      placeholder="John Doe"
                      required={true}
                    >
                      <input
                        id="username"
                        maxLength={30}
                        onBlur={[Function]}
                        onChange={[Function]}
                        placeholder="John Doe"
                        required={true}
                        value=""
                      />
                    </t>
                  </r>
                  <label
                    htmlFor="username"
                  >
                    UserName
                  </label>
                </div>
                <div
                  className="tms-textfield"
                >
                  <r
                    field="password"
                    id="password"
                    maxLength={30}
                    placeholder="Your password"
                    required={true}
                    type="password"
                  >
                    <t
                      fieldApi={
                        Object {
                          "doneValidatingField": [Function],
                          "getError": [Function],
                          "getFieldName": [Function],
                          "getSuccess": [Function],
                          "getTouched": [Function],
                          "getValue": [Function],
                          "getWarning": [Function],
                          "registerAsyncValidation": [Function],
                          "reset": [Function],
                          "setError": [Function],
                          "setSuccess": [Function],
                          "setTouched": [Function],
                          "setValue": [Function],
                          "setWarning": [Function],
                          "submits": 1,
                          "submitted": false,
                          "validatingField": [Function],
                        }
                      }
                      id="password"
                      maxLength={30}
                      placeholder="Your password"
                      required={true}
                      type="password"
                    >
                      <input
                        id="password"
                        maxLength={30}
                        onBlur={[Function]}
                        onChange={[Function]}
                        placeholder="Your password"
                        required={true}
                        type="password"
                        value=""
                      />
                    </t>
                  </r>
                  <label
                    htmlFor="password"
                  >
                    Password
                  </label>
                </div>
                <button
                  className="tms-btn tms-btn--primary tms-btn--block"
                >
                  SIGN IN
                </button>
              </form>
            </div>
          </t>
        </Connect(t)>
      </t>
    </div>
  </div>
</LoginPage>
`;
